File: tarai.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <numeric>
#include <cmath>
#include <version>

using namespace std;

int tarai(int x, int y, int z) {
    if (x <= y) {
        return y;
    } else {
        return tarai(tarai(x - 1, y, z), tarai(y - 1, z, x), tarai(z - 1, x, y));
    }
}

double average(const vector<double>& v) {
    return reduce(v.begin(), v.end()) / v.size();
}

double standard_deviation(const vector<double>& v) {
    auto avg = average(v);
    return sqrt(transform_reduce(v.begin(), v.end(), 0.0, plus<>(), [avg](double x) { return (x - avg) * (x - avg); }) / v.size());
}

int main() {
    const int n_iter = 10;
    vector<double> times;

    for (int i = 0; i < n_iter; i++) {
        auto start_time = chrono::high_resolution_clock::now();
        tarai(13, 5, 0);
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        times.push_back(elapsed_time.count());
    }

    auto avg_time = average(times);
    auto std_dev_time = standard_deviation(times);

    cout << "C++ version: " << __cplusplus << "\n";
    cout << "Average elapsed_time: " << avg_time << " sec\n";
    cout << "Standard deviation: " << std_dev_time << " sec\n";

    return 0;
}

File: tarai.java
import java.time.Duration;
import java.time.Instant;

public class tarai {
    public static int tarai(int x, int y, int z) {
        if (x <= y) {
            return y;
        } else {
            return tarai(tarai(x - 1, y, z),
                         tarai(y - 1, z, x),
                         tarai(z - 1, x, y));
        }
    }

    public static void main(String[] args) {
        int nIter = 10;
        long[] times = new long[nIter];

        for (int i = 0; i < nIter; i++) {
            Instant start = Instant.now();
            tarai(13, 5, 0);
            Instant end = Instant.now();
            Duration duration = Duration.between(start, end);
            times[i] = duration.toMillis();
        }

        double averageTime = 0, stdTime = 0;
        for (long time : times) {
            averageTime += time;
        }
        averageTime /= nIter;

        for (long time : times) {
            stdTime += Math.pow(time - averageTime, 2);
        }
        stdTime /= nIter;
        stdTime = Math.sqrt(stdTime);

        System.out.println("Java version: " + System.getProperty("java.version"));
        System.out.println("Average elapsed_time: " + averageTime/1000 + " sec");
        System.out.println("Standard deviation: " + stdTime/1000 + " sec");
    }
}
